src/util.rs

```diff
@@ -1,47 +1,54 @@
 const HTTP_DATE_FMT: &str = "%a, %d %b %Y %H:%M:%S";
 
 /// Size of a file on disk. 0 if it doesn't exist.
+#[must_use]
 pub fn file_size(path: &str) -> usize {
     std::fs::File::open(path)
         .map(|f| f.metadata().map(|m| m.len()).unwrap_or(0))
         .unwrap_or(0) as usize
 }
 
 /// Does what it says.
+#[must_use]
 pub fn decode_form_value(post: &str) -> String {
     let cleaned = post.replace('+', " ").replace('\r', "");
     percent_decode(&cleaned).unwrap_or_else(|| "".into())
 }
 
 /// Current date in HTTP format.
+#[must_use]
 pub fn http_current_date() -> String {
     let now = libc_strftime::epoch();
     libc_strftime::strftime_gmt(HTTP_DATE_FMT, now) + " GMT"
 }
 
 /// Mutably borrowed from the zero dependency httpserv project.
-/// https://github.com/nic-hartley/httpserv/blob/585c020/src/http.rs
+/// <https://github.com/nic-hartley/httpserv/blob/585c020/src/http.rs>
+#[must_use]
 pub fn percent_decode(mut inp: &str) -> Option<String> {
     let mut out = Vec::new();
-    loop {
+    for _ in 0..512 {
         let next_pct = match inp.find('%') {
-            Some(l) if l < inp.len() - 2 => l,
+            Some(l) if l < if inp.len() > 1 { inp.len() - 2 } else { 0 } => l,
```

Comment:
Bit of a botchy fix for the header issue but it caused thread panics so I quickly fixed it